main.go

```diff
@@ -208,6 +211,52 @@ func (d *Downloader) chunks(t uint64) []chunk {
 	return c
 }
 
+func (d *Downloader) prepareAndStartDownload(ctx context.Context, url string, wg *sync.WaitGroup, ch chan<- DownloadStatus) {
+	defer wg.Done()
+	path := filepath.Join(os.TempDir(), filepath.Base(url))
+	s := DownloadStatus{URL: url, DownloadedFilePath: path}
+	t, err := d.getDownloadSize(ctx, url) // TODO: retry
+	if err != nil {
+		s.Error = fmt.Errorf("error getting file size: %w", err)
+		ch <- s
+		return
+	}
+	s.FileSizeBytes = t
+	ch <- s // send total file size to the user
+	f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		s.Error = fmt.Errorf("error creating %s: %w", path, err)
+		ch <- s
+		return
+	}
+	defer f.Close()
+	if err := f.Truncate(int64(t)); err != nil {
+		s.Error = fmt.Errorf("error truncating %s to %d: %w", path, t, err)
+		ch <- s
+		return
+	}
+	for _, c := range d.chunks(t) {
+		b, err := d.downloadChunk(ctx, url, c)
+		if err != nil {
+			s.Error = err
+			ch <- s
+			return
+		}
+		_, err = f.WriteAt(b, int64(c.start))
+		if err != nil {
+			s.Error = fmt.Errorf("error writing to %s: %w", path, err)
+			ch <- s
+			return
+		}
+		a := int64(len(b))
+		if !c.last() {
+			a -= 1 // adjust for extra EOF bytes
+		}
+		atomic.AddInt64(&s.DownloadedFileBytes, a)
```

Comment:
Each goroutine will access a different chunk, right? So, the usage of atomic does not seem to be needed

Reply:
Right now it is. But next we're might make chunks be downloaded in parallel. Gonna clean up and we change later if needed : )